File: unid-node-wallet-sdk/src/core/interfaces/did-operation.ts
import { DIDDocument } from "./did-document";
import { KeyPairSecp256K1, PublicKeyType } from "./key-pair";

export interface PublicKeyPayload {
    id  : string,
    type: PublicKeyType,
    jwk : KeyPairSecp256K1,
    purpose: Array<string>,
}

// ACTION: add-public-keys
export interface DIDAddPublicKeysPayload extends PublicKeyPayload {
}

export interface DIDAddPublicKeysAction {
    action     : 'add-public-keys',
    public_keys: Array<DIDAddPublicKeysPayload>
}

// ACTION: remove-public-keys
export interface DIDRemovePublicKeysAction {
    action: 'remove-public-keys',
    ids   : Array<string>,
}

// ACTION: add-services
export interface DIDAddServicesPayload {
}

export interface DIDAddServicesAction {
    action  : 'add-services',
    services: Array<DIDAddServicesPayload>,
}

// ACTION: remove-services
export interface DIDRemoveServicesAction {
    action: 'remove-services',
    ids   : Array<string>,
}

// ACTION: replace
export interface DIDReplacePayload {
    public_keys: Array<PublicKeyPayload>,
    service_endpoints: Array<string>,
}

export interface DIDReplaceAction {
    action  : 'replace',
    document: DIDReplacePayload,
}

export interface DIDReplaceDeltaObject {
    patches: Array<DIDReplaceAction>,
    update_commitment: string,
}

export interface DIDReplaceSuffixObject {
    delta_hash: string,
    recovery_commitment: string,
}

// ACTION: ietf-json-patch
export interface DIDIetfJsonPatchAction {
    action : 'replace',
    patches: Array<any>
}

export interface DIDResolutionRequest {
    did: string
}

export interface DIDResolutionResponse {
    '@context': 'https://www.w3.org/ns/did-resolution/v1',
    didDocument: DIDDocument,
    methodMetadata: {
        published: boolean,
        recoveryCommitment: string,
        updateCommitment: string,
    },
}

export interface DIDCreateRequest {
    publicKeys: Array<PublicKeyPayload>,
    commitmentKeys: {
        recovery: KeyPairSecp256K1,
        update  : KeyPairSecp256K1,
    },
    serviceEndpoints: Array<string>,
}

export interface DIDCreatePayload {
    type: 'create',
    delta: string,
    suffix_data: string,
}

export interface DIDCreateResponse {
    '@context': 'https://www.w3.org/ns/did-resolution/v1',
    didDocument: DIDDocument,
    methodMetadata: {
        published: boolean,
        recoveryCommitment: string,
        updateCommitment: string,
    },
}

export interface DIDUpdateRequest {
    // NOT IMPLEMENTED
}

export interface DIDUpdateResponse {
    // NOT IMPLEMENTED
}

export interface DIDRecoverRequest {
    // NOT IMPLEMENTED
}

export interface DIDRecoverResponse {
    // NOT IMPLEMENTED
}

export interface DIDDeactivateRequest {
    // NOT IMPLEMENTED
}

export interface DIDDeactivateResponse {
    // NOT IMPLEMENTED
}

File: src/unid/did/interfaces/did_operation.rs
use super::did_document::*;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DIDCreateRequest {
    pub public_keys: Vec<PublicKeyPayload>,
    pub commitment_keys: CommitmentKeys,
    pub service_endpoints: Vec<String>,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CommitmentKeys {
    pub recovery: KeyPairSecp256K1,
    pub update: KeyPairSecp256K1,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PublicKeyPayload {
    pub id: String,
    #[serde(rename = "type")]
    pub type_field: String,
    pub jwk: KeyPairSecp256K1,
    pub purpose: Vec<String>,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DIDCreatePayload {
    #[serde(rename = "type")]
    pub type_field: String,
    pub delta: String,
    pub suffix_data: String,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct DIDReplacePayload {
    pub public_keys: Vec<PublicKeyPayload>,
    pub service_endpoints: Vec<String>,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct DIDReplaceAction {
    pub action: String,
    pub document: DIDReplacePayload,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct DIDReplaceDeltaObject {
    pub patches: Vec<DIDReplaceAction>,
    pub update_commitment: String,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct DIDReplaceSuffixObject {
    pub delta_hash: String,
    pub recovery_commitment: String,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct DIDResolutionRequest {
    pub did: String,
}

pub type KV = HashMap<String, String>;